hw3/season06/_matrix.cpp

```diff
@@ -0,0 +1,253 @@
+#include <iostream>
+#include <vector>
+#include <stdexcept>
+#include <mkl.h>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/numpy.h>
+#include <pybind11/operators.h>
+
+using namespace std;
+namespace py = pybind11;
+
+class Matrix
+{
+public:
+    // default constructor
+    Matrix() : _rows(0), _cols(0) {};
+
+    Matrix(size_t row, size_t col) : _rows(row), _cols(col)
+    {
+        _vec.clear();
+        _vec.resize(row * col, 0);
+    }
+
+    // copy constructor
+    Matrix(Matrix const &other) : _rows(other._rows), _cols(other._cols)
+    {
+        _vec.resize(_rows, _cols);
+        for (size_t i = 0; i < _rows; i++)
+            for (size_t j = 0; j < _cols; j++)
+                (*this)(i, j) = other(i, j);
+    }
+
+    // move constructor
+    Matrix(Matrix &&other) : _rows(other._rows), _cols(other._cols)
+    {
+        this->_vec = std::move(other._vec);
+    }
+
+    // copy assignment operator
+    Matrix &operator=(Matrix const &other)
+    {
+        if (this != &other)
+        {
+            this->_cols = other._cols;
+            this->_rows = other._rows;
+            _vec.resize(_rows, _cols);
+            for (size_t i = 0; i < _rows; i++)
+                for (size_t j = 0; j < _cols; j++)
+                    (*this)(i, j) = other(i, j);
+        }
+        return *this;
+    }
+
+    // move assignment operator
+    Matrix &operator=(Matrix &&other)
+    {
+        if (this != &other)
+        {
+            this->_cols = other._cols;
+            this->_rows = other._rows;
+            this->_vec = std::move(other._vec);
+        }
+        return *this;
+    }
+
+    // destructor
+    ~Matrix()
+    {
+        _vec.clear(); 
+        _vec.shrink_to_fit();
+    }
+
+    // getter
+    const size_t nrow() const
+    {
+        return _rows;
+    }
+    const size_t ncol() const
+    {
+        return _cols;
+    }
+    double *addr()
+    {
+        return &_vec[0];
+    }
+
+    // load numpy array to matrix
+    void load_from_array(py::array_t<double> array)
```

Comment:
This mixes the (pure) C++ implementation with the wrapper (pybind11).  It's ok to do it in homework, but a bad idea in production code, because it requires you to include Python.h no matter you use Python or not.

One way to separate the C++ kernel from the wrapper can be found at: 
https://github.com/solvcon/modmesh/blob/68681c7111beb6eaf8489bf1a6ed77aa0167986e/include/modmesh/python/common.hpp#L196